service/frontend/adminHandler.go

```diff
@@ -421,6 +423,237 @@ func (adh *AdminHandler) GetWorkflowExecutionRawHistory(
+	)
+	startEventID, endEventID, versionHistory, err := adh.getEventRange(
+		request,
+		versionHistories,
+	)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	token, err := adh.preparePaginationToken(
+		ctx,
+		request,
+		versionHistory.GetBranchToken(),
+		startEventID,
+		endEventID)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	pageSize := int(request.GetMaximumPageSize())
+	shardID := common.WorkflowIDToHistoryShard(
+		execution.GetWorkflowId(),
+		adh.numberOfHistoryShards,
+	)
+	rawHistoryResponse, err := adh.historyV2Mgr.ReadRawHistoryBranch(&persistence.ReadHistoryBranchRequest{
+		BranchToken:   token.BranchToken,
+		MinEventID:    startEventID,
+		MaxEventID:    endEventID,
+		PageSize:      pageSize,
+		NextPageToken: token.PersistenceToken,
+		ShardID:       common.IntPtr(shardID),
+	})
+	if err != nil {
+		if _, ok := err.(*gen.EntityNotExistsError); ok {
+			// when no events can be returned from DB, DB layer will return
+			// EntityNotExistsError, this API shall return empty response
+			return &admin.GetWorkflowExecutionRawHistoryV2Response{
+				HistoryBatches: []*gen.DataBlob{},
+				NextPageToken:  nil, // no further pagination
+				VersionHistory: versionHistory.ToThrift(),
+			}, nil
+		}
+		return nil, err
+	}
+
+	token.PersistenceToken = rawHistoryResponse.NextPageToken
+	size := rawHistoryResponse.Size
+	// N.B. - Dual emit is required here so that we can see aggregate timer stats across all
+	// domains along with the individual domains stats
+	adh.metricsClient.RecordTimer(scope, metrics.HistorySize, time.Duration(size))
+	domainScope := adh.metricsClient.Scope(scope, metrics.DomainTag(request.GetDomain()))
+	domainScope.RecordTimer(metrics.HistorySize, time.Duration(size))
+
+	rawBlobs := rawHistoryResponse.HistoryEventBlobs
+	blobs := []*gen.DataBlob{}
+	for _, blob := range rawBlobs {
+		blobs = append(blobs, blob.ToThrift())
+	}
+
+	result := &admin.GetWorkflowExecutionRawHistoryV2Response{
+		HistoryBatches: blobs,
+		VersionHistory: versionHistory.ToThrift(),
+	}
+	if len(token.PersistenceToken) == 0 {
+		result.NextPageToken = nil
+	} else {
+		result.NextPageToken, err = serializeHistoryToken(token)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+func (adh *AdminHandler) validateGetWorkflowExecutionRawHistoryV2Request(
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+) error {
+
+	execution := request.Execution
+	if len(execution.GetWorkflowId()) == 0 {
+		return &gen.BadRequestError{Message: "Invalid WorkflowID."}
+	}
+	// TODO currently, this API is only going to be used by re-send history events
+	// to remote cluster if kafka is lossy again, in the future, this API can be used
+	// by CLI and client, then empty runID (meaning the current workflow) should be allowed
+	if len(execution.GetRunId()) == 0 || uuid.Parse(execution.GetRunId()) == nil {
+		return &gen.BadRequestError{Message: "Invalid RunID."}
+	}
+
```

Comment:
there should also be a check on both start event ID / version & end event ID / version missing

Reply:
updated